Improve generated dump debugging instructions (#46493)
dotnet/runtime: eng/testing/debug-dump-template.md

# Get the dump

Click the link to the dump on the `Helix Test Logs` tab in Azure DevOps. This is the same place you got these instructions from.

# Get the Helix payload

[Runfo](https://github.com/jaredpar/runfo/tree/master/runfo#runfo) helps get information about helix test runs and azure devops builds. We will use it to download the payload and symbols:

> ```script

Reviewer: nit: only this file in the repo is using ` ```script `, which is not a [known language alias](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml). In other documents, ` ```sh ` and ` ```cmd ` are used.

Reviewer 2: According to that file, `cmd` isn't either! I'll use `sh` which will at least be right for Unix. it doesn't really matter.

Reviewer: cmd definitely is.
https://github.com/github/linguist/blob/8fc4898d34dbe9323bc1bb7971652998d3d39b4f/lib/linguist/languages.yml#L449

```cmd
:: this is cmd

setx FOO=BAR
echo %FOO%
```

Reviewer: Extensions are also included in language name, so cmd is valid https://github.com/dotnet/runtime/pull/46493#discussion_r550953706.

dotnet tool install --global runfo
dotnet tool update --global runfo
```
If prompted, open a new command prompt to pick up the updated PATH.

> ```script
> # On Windows
> runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %WOUTDIR%
> # On Linux and macOS
> runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %LOUTDIR%

Reviewer: `$JOBID` instead of `%JOBID%` on Linux/macOS?
````suggestion
```cmd
:: On Windows
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %WOUTDIR%
```
```sh
# On Linux and macOS
runfo get-helix-payload -j $JOBID -w $WORKITEM -o $LOUTDIR
````

Reviewer 2: Thanks, but they are'nt environment variables
https://github.com/dotnet/runtime/pull/46493/files#diff-fd8bb0da491f288f6e268f835a49b14126ba3f1a15ab6971ddc76dd480b81b11R97

# Get the dump

Click the link to the dump on the `Helix Test Logs` tab in Azure DevOps. This is the same place you got these instructions from.

# Get the Helix payload

[Runfo](https://github.com/jaredpar/runfo/tree/master/runfo#runfo) helps get information about helix test runs and azure devops builds. We will use it to download the payload and symbols:

> sos

Reviewer 3: ```suggestion
```

# On Linux and macOS
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %LOUTDIR%
```

> NOTE: if the helix job is an internal job, you need to pass down a [helix authentication token](https://helix.dot.net/Account/Tokens) using the `--helix-token` argument.

Now extract the files:

```script
# On Windows
for /f %i in ('dir /s/b %WOUTDIR%\*zip') do tar -xf %i -C %WOUTDIR%

# On Linux and macOS
# obtain `unzip` if necessary; eg `sudo apt-get install unzip` or `sudo dnf install unzip`
find %LOUTDIR% -name '*zip' -exec unzip -d %LOUTDIR% {} \;
```

Now use the [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos) to install the SOS debugging extension.
```script
dotnet tool install --global dotnet-sos
dotnet tool update --global dotnet-sos
```
If prompted, open a new command prompt to pick up the updated PATH.
```script
dotnet sos install --architecture Arm
dotnet sos install --architecture Arm64
dotnet sos install --architecture x86
dotnet sos install --architecture x64
```

# Now choose a section below based on your OS.

## If it's a Windows dump on Windows...

## ... and you want to debug with WinDbg

> Install or update WinDbg if necessary ([external](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools), [internal](https://osgwiki.com/wiki/Installing_WinDbg)). If you don't have a recent WinDbg you may have to do `.update sos`.

Reviewer 3: Is there any advantage in using MS internal link to downloading windbg?

Reviewer 2: Sometimes it has fixes that aren't on the stable release bits.

---
## If it's a Linux dump on Linux...

## ... and you want to debug with LLDB

Install or update LLDB if necessary ([instructions here](https://github.com/dotnet/diagnostics/blob/master/documentation/lldb/linux-instructions.md))

Load the dump:
```script
lldb --core <path-to-dmp> %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0/dotnet
```

Within lldb:
```script
setclrpath %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
sethostruntime /usr/bin/dotnet
setsymbolserver -directory %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
```
If you want to load native symbols
```
loadsymbols
```

## ... and you want to debug with dotnet-dump

Install the [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
```script
dotnet tool install --global dotnet-dump
dotnet tool update --global dotnet-dump
```
If prompted, open a new command prompt to pick up the updated PATH.
```script
dotnet-dump analyze <path-to-dump>
```
Within dotnet-dump:
```script
setclrpath %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
setsymbolserver -directory %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
```

---

> ## If it's a MacOS dump

Reviewer 3: ```suggestion
## If it's a macOS dump
```

> Instructions for debugging dumps on MacOS are essentially the same as [Linux](#If-it's-a-Linux-dump-on-Linux...) with one exception: `dotnet-dump` cannot analyze macOS system dumps: you must use `lldb` for those. `dotnet-dump` can only analyze dumps created by `dotnet-dump` or `createdump` or by the runtime on hangs.

Reviewer 3: ```suggestion
Instructions for debugging dumps on macOS are essentially the same as [Linux](#If-it's-a-Linux-dump-on-Linux...) with one exception: `dotnet-dump` cannot analyze macOS system dumps: you must use `lldb` for those. `dotnet-dump` can only analyze dumps created by `dotnet-dump` or `createdump` or by the runtime on hangs.
```

```sh
dotnet tool install --global runfo
dotnet tool update --global runfo
```
If prompted, open a new command prompt to pick up the updated PATH.
```sh
# On Windows
# assumes %WOUTDIR% does not exist
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %WOUTDIR%
# On Linux and macOS
# assumes %LOUTDIR% does not exist
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o %LOUTDIR%
```

> NOTE: if the helix job is an internal job, you need to pass down a [helix authentication token](https://helix.dot.net/Account/Tokens) using the `--helix-token` argument.

Now extract the files:

```sh
# On Windows
for /f %i in ('dir /s/b %WOUTDIR%\*zip') do tar -xf %i -C %WOUTDIR%

# On Linux and macOS
# obtain `unzip` if necessary; eg `sudo apt-get install unzip` or `sudo dnf install unzip`
find %LOUTDIR% -name '*zip' -exec unzip -d %LOUTDIR% {} \;
```

Now use the [dotnet-sos global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-sos) to install the SOS debugging extension.
```sh
dotnet tool install --global dotnet-sos
dotnet tool update --global dotnet-sos
```
If prompted, open a new command prompt to pick up the updated PATH.
```sh

> dotnet sos install --architecture Arm

Reviewer 4: The last one will override the others, at least on Linux (not sure about windows, haven't seen if it is conditional on the bitness of the extension host, but given the errors I've seen I'd say no). Users should only install the one they are going to use.

Reviewer 2: Oh - I didn't notice that because I was testing with x64, which is last!

Could we imagine installing them all, and picking the right one - eliminate another decision point? cc @mikem8361

Reviewer 2: I'll fix the text meantime

Reviewer 2: It looks like it defaults silently to the bitness of the dotnet.exe you're running it with (?)

Reviewer 5: Yes it does default to the architecture of the dotnet.exe you are running (the most common case).

As far as installing all the of the architectures, there would have to be a rid named subdirectory to separate them so the decision point would be postponed a little to the .load command in windbg. 

This manual dotnet-sos install/load on Windows was really meant as a fallback while we wait for the public debugger to include SOS (all architectures) like the internal one does now.  The internal Windows debugger automatically loads the correct architecture and latest version of SOS from the extension gallery.

Reviewer 2: Oh perfect then it will go away.

---
## If it's a Linux dump on Linux...

## ... and you want to debug with LLDB

Install or update LLDB if necessary ([instructions here](https://github.com/dotnet/diagnostics/blob/master/documentation/lldb/linux-instructions.md))

Load the dump:
```sh
lldb --core <path-to-dmp> %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0/dotnet
```

Within lldb:
```
setclrpath %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
sethostruntime /usr/bin/dotnet
setsymbolserver -directory %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
```
If you want to load native symbols
```
loadsymbols
```

## ... and you want to debug with dotnet-dump

Install the [dotnet-dump global tool](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump).
```sh
dotnet tool install --global dotnet-dump
dotnet tool update --global dotnet-dump
```
If prompted, open a new command prompt to pick up the updated PATH.
```sh
dotnet-dump analyze <path-to-dump>
```
Within dotnet-dump:
```sh
setclrpath %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
setsymbolserver -directory %LOUTDIR%/shared/Microsoft.NETCore.App/6.0.0
```

---
## If it's a macOS dump

> Instructions for debugging dumps on macOS are essentially the same as [Linux](#If-it's-a-Linux-dump-on-Linux...) with one exception: `dotnet-dump` cannot analyze macOS system dumps: you must use `lldb` for those. `dotnet-dump` can only analyze dumps created by `dotnet-dump` or `createdump` or by the runtime on hangs.

Reviewer 4: ```suggestion
Instructions for debugging dumps on macOS are essentially the same as [Linux](#If-it's-a-Linux-dump-on-Linux...) with one exception: `dotnet-dump` cannot analyze macOS system dumps: you must use `lldb` for those. `dotnet-dump` can only analyze dumps created by `dotnet-dump` or `createdump`, by the runtime on crashes when the appropriate environment variables are set, or the [`blame-hang` setting of `dotnet test`](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-test).
```
